h/formatters/annotation_moderation.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+import sqlalchemy as sa
+from zope.interface import implementer
+
+from h import models
+from h.formatters.interfaces import IAnnotationFormatter
+
+
+@implementer(IAnnotationFormatter)
+class AnnotationModerationFormatter(object):
+    """
+    Formatter for exposing an annotation's moderation information.
+
+    If the passed-in user has permission to hide the annotation (if they are a
+    moderator of the annotation's group, for instance), this formatter will
+    add a `moderation` key to the payload, with a count of how many users have
+    flagged the annotation.
+    """
+
+    def __init__(self, session, user, has_permission):
+        self._session = session
+        self._user = user
+        self._has_permission = has_permission
+
+        # Local cache of flag counts. We don't need to care about detached
+        # instances because we only store the annotation id and a count.
+        self._cache = {}
+
+    def preload(self, ids):
+        if self._user is None:
+            return
+
+        query = self._session.query(sa.func.count(models.Flag.id).label('flag_count'),
+                                    models.Flag.annotation_id) \
+                             .filter(models.Flag.annotation_id.in_(ids)) \
+                             .group_by(models.Flag.annotation_id)
+
+        flag_counts = {f.annotation_id: f.flag_count for f in query}
+        missing_ids = set(ids) - set(flag_counts.keys())
```

Comment:
Again, certainly not a blocker for merging this, but it might be slightly less work to do this the other way round. That is:

```python
flag_counts = {id_: 0 for id_ in ids}
flag_counts.update({f.annotation_id: f.flag_count for f in query})
```